### PR TITLE
More general URL handling

### PR DESCRIPTION
**Pull request by LeaVerou**
- Backends can define URL criteria in which case they don't need to define a `test()` function, the base one Just Works™
- Backends can define URL regex patterns, which get parsed automatically. We could also use `URLPattern` in the future.
- For backends with a singular `host`, users can now use relative URLs with backends (useful for when they are not using `Backend.from()` but specifying the backend directly. Maybe we should extend this to backends with multiple hosts and just use the first one.

### CODE REVIEW

File: src/util.js [hunk 1]
```diff
@@ -82,4 +82,46 @@ export function phrase (me, id, ...args) {
...
+ * @param {string | URL} url
+ * @param {{protocol, host, urls} | {protocol, host, urls}[]} criteria
+ * @returns {boolean}
+ */
+export function testURL (url, criteria) {
+	if (typeof url === "string") {
+		url = new URL(url);
+	}
+
+	if (Array.isArray(criteria.urls)) {
```
Comment (DmitrySharabin) [inline src/util.js:98]:
```suggestion
	if (Array.isArray(criteria)) {
```

File: src/util.js [hunk 2]
```diff
@@ -82,4 +82,46 @@ export function phrase (me, id, ...args) {
...
+		return criteria.urls.some(pattern => testURL(url, pattern));
```
Comment (DmitrySharabin) [inline src/util.js:99]:
```suggestion
		return criteria.some(pattern => testURL(url, pattern));
```

**Review by DmitrySharabin — CHANGES_REQUESTED** [review]
This is so clever! 🤩

Please see some comments (mostly questions) and a couple of suggestions.

File: src/util.js (hunk 1 above)
Comment (DmitrySharabin) [reply src/util.js:98]: When we have `urls`, we pass them as `criteria` in the `test()` method. So, we already have the needed array here.

File: src/util.js [hunk 3]
```diff
@@ -82,4 +82,46 @@ export function phrase (me, id, ...args) {
...
+	let {protocol, host, path} = criteria;
+
+	if (protocol && url.protocol != protocol) {
+		return false;
+	}
+
+	if (host) {
+		if (host.startsWith("*.")) {
+			// Wildcard subdomain
+			host = host.replace(/^(\*\.)+/, "");
```
[the remark this answers is not in the document]
Comment (LeaVerou) [reply src/util.js:111]: It’s a balance between scoping it down enough and anticipating future use cases on one hand and not complicating it too much on the other. Right now we only have cases where there is one asterisk and only at the start. I guess the way it’s defined `*.*.` would be pointless though so I’m happy to remove the +

File: src/util.js [hunk 4]
```diff
@@ -82,4 +82,46 @@ export function phrase (me, id, ...args) {
 	// We're on the root class and still can't find it
 	// Fall back to just displaying the id and all the args right after it
 	return id + " " + args.join(" ");
+}
+
+/**
+ * Test whether a URL matches a set of criteria
+ * @param {string | URL} url
+ * @param {{protocol, host, urls} | {protocol, host, urls}[]} criteria
```
[the remark this answers is not in the document]
Comment (LeaVerou) [reply src/util.js:90]: We do, or at least we should! That’s the `Array.isArray()` part!

File: src/util.js [hunk 5]
```diff
@@ -82,4 +82,46 @@ export function phrase (me, id, ...args) {
...
+		url = new URL(url);
+	}
+
+	if (Array.isArray(criteria.urls)) {
+		return criteria.urls.some(pattern => testURL(url, pattern));
+	}
+
+	let {protocol, host, path} = criteria;
+
+	if (protocol && url.protocol != protocol) {
```
[the remark this answers is not in the document]
Comment (LeaVerou) [reply src/util.js:104]: Nope, just oversight!

```suggestion
	if (protocol && url.protocol !== protocol) {
```

File: src/util.js [hunk 6]
```diff
@@ -82,4 +82,46 @@ export function phrase (me, id, ...args) {
...
+	if (host) {
+		if (host.startsWith("*.")) {
+			// Wildcard subdomain
+			host = host.replace(/^(\*\.)+/, "");
+
+			if (url.host !== host && !url.host.endsWith("." + host)) {
+				return false;
+			}
+		}
+		else if (url.host != host) {
```
[the remark this answers is not in the document]
Comment (LeaVerou) [reply src/util.js:117]:
```suggestion
		else if (url.host !== host) {
```

File: backends/firebase/index.js [hunk 7]
```diff
@@ -272,20 +272,18 @@ export default class Firebase extends AuthBackend {
 		path: "madata/data"
 	};
 
-	static test (url) {
-		url = new URL(url);
-		return url.host.endsWith("firebaseio.com");
-	}
+	static urls = [
+		{ host: "*.firebaseio.com" },
+		{ host: "firebasestorage.googleapis.com" },
```
[the remark this answers is not in the document]
Comment (LeaVerou) [reply backends/firebase/index.js:277]: I thought of this but it appears that protocol is used more broadly for Google Cloud stuff?

[the remark this answers is not in the document]
Comment (DmitrySharabin) [reply backends/firebase/index.js:277]: It is. In the Firebase backend, it's used only when we delete a file from the storage. I was also thinking of allowing authors to load files from the storage. But it seems to me a bit out of Madata's scope. No? Let's leave it as it is for now. We can add it later if needed.

File: src/util.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (DmitrySharabin) [reply src/util.js:90]: I can see it now after we replaced `criteria.urls` with `criteria` in the function body. :)

**Review by DmitrySharabin — APPROVED** [review]

**Pull request merged**